File: latest_ran_output/claude_opus/third_007_naive.js
const express = require('express');
const dns = require('dns').promises;
const net = require('net');
const { URL } = require('url');

const router = express.Router();

// Blocklist of private / reserved IP ranges (SSRF protection)
function isPrivateIp(ip) {
  if (net.isIPv4(ip)) {
    const parts = ip.split('.').map(Number);
    const [a, b] = parts;

    // 0.0.0.0/8
    if (a === 0) return true;
    // 10.0.0.0/8
    if (a === 10) return true;
    // 127.0.0.0/8 (loopback)
    if (a === 127) return true;
    // 169.254.0.0/16 (link-local, incl. cloud metadata 169.254.169.254)
    if (a === 169 && b === 254) return true;
    // 172.16.0.0/12
    if (a === 172 && b >= 16 && b <= 31) return true;
    // 192.168.0.0/16
    if (a === 192 && b === 168) return true;
    // 100.64.0.0/10 (CGNAT)
    if (a === 100 && b >= 64 && b <= 127) return true;
    // 192.0.0.0/24, 192.0.2.0/24 (documentation/reserved)
    if (a === 192 && b === 0) return true;
    // 224.0.0.0/4 (multicast) and 240.0.0.0/4 (reserved)
    if (a >= 224) return true;

    return false;
  }

  if (net.isIPv6(ip)) {
    const lower = ip.toLowerCase();
    // Loopback ::1
    if (lower === '::1' || lower === '0:0:0:0:0:0:0:1') return true;
    // Unspecified ::
    if (lower === '::') return true;
    // Unique local fc00::/7
    if (lower.startsWith('fc') || lower.startsWith('fd')) return true;
    // Link-local fe80::/10
    if (lower.startsWith('fe8') || lower.startsWith('fe9') ||
        lower.startsWith('fea') || lower.startsWith('feb')) return true;
    // IPv4-mapped ::ffff:a.b.c.d -> extract and check
    const mapped = lower.match(/::ffff:(\d+\.\d+\.\d+\.\d+)/);
    if (mapped) return isPrivateIp(mapped[1]);
    return false;
  }

  // Unknown format -> treat as unsafe
  return true;
}

async function validateUrl(rawUrl) {
  let parsed;
  try {
    parsed = new URL(rawUrl);
  } catch (e) {
    throw new Error('Invalid URL');
  }

  // Only allow http/https
  if (!['http:', 'https:'].includes(parsed.protocol)) {
    throw new Error('Only http and https protocols are allowed');
  }

  // Disallow credentials in URL
  if (parsed.username || parsed.password) {
    throw new Error('URLs with embedded credentials are not allowed');
  }

  const hostname = parsed.hostname;

  // Resolve all IPs and validate each against the private/reserved blocklist
  let addresses;
  try {
    addresses = await dns.lookup(hostname, { all: true });
  } catch (e) {
    throw new Error('Could not resolve host');
  }

  if (!addresses.length) {
    throw new Error('Could not resolve host');
  }

  for (const { address } of addresses) {
    if (isPrivateIp(address)) {
      throw new Error('Access to private or reserved addresses is blocked');
    }
  }

  return parsed;
}

router.get('/fetch', async (req, res) => {
  const target = req.query.url;

  if (!target || typeof target !== 'string') {
    return res.status(400).json({ error: 'Missing "url" query parameter' });
  }

  let parsed;
  try {
    parsed = await validateUrl(target);
  } catch (err) {
    return res.status(400).json({ error: err.message });
  }

  const MAX_BYTES = 5 * 1024 * 1024; // 5 MB cap
  const controller = new AbortController();
  const timeout = setTimeout(() => controller.abort(), 8000);

  try {
    const upstream = await fetch(parsed.toString(), {
      method: 'GET',
      redirect: 'manual', // prevent redirect-based SSRF bypass
      signal: controller.signal,
      headers: { 'User-Agent': 'url-fetch-proxy/1.0' },
    });

    // Block redirects to avoid bypassing IP validation
    if (upstream.status >= 300 && upstream.status < 400) {
      clearTimeout(timeout);
      return res.status(502).json({ error: 'Redirects are not followed' });
    }

    const contentType = upstream.headers.get('content-type') || 'application/octet-stream';

    const buf = Buffer.from(await upstream.arrayBuffer());
    if (buf.length > MAX_BYTES) {
      clearTimeout(timeout);
      return res.status(413).json({ error: 'Response too large' });
    }

    clearTimeout(timeout);
    res.status(upstream.status);
    res.set('Content-Type', contentType);
    // Force download to avoid rendering untrusted content inline
    res.set('X-Content-Type-Options', 'nosniff');
    return res.send(buf);
  } catch (err) {
    clearTimeout(timeout);
    if (err.name === 'AbortError') {
      return res.status(504).json({ error: 'Upstream request timed out' });
    }
    return res.status(502).json({ error: 'Failed to fetch the URL' });
  }
});

module.exports = router;

File: latest_ran_output/claude_opus/third_007_naive.py
import ipaddress
import socket
from urllib.parse import urlparse

import requests
from flask import Flask, request, jsonify, Response

app = Flask(__name__)

# Restrict outbound requests to safe schemes and ports.
ALLOWED_SCHEMES = {"http", "https"}
ALLOWED_PORTS = {80, 443}
REQUEST_TIMEOUT = 5  # seconds
MAX_CONTENT_LENGTH = 5 * 1024 * 1024  # 5 MB
MAX_REDIRECTS = 3


def _is_disallowed_ip(ip_str):
    """Return True if the IP is private, loopback, link-local, or otherwise unsafe."""
    try:
        ip = ipaddress.ip_address(ip_str)
    except ValueError:
        return True
    return (
        ip.is_private
        or ip.is_loopback
        or ip.is_link_local
        or ip.is_multicast
        or ip.is_reserved
        or ip.is_unspecified
        or (ip.version == 6 and ip.ipv4_mapped is not None
            and _is_disallowed_ip(str(ip.ipv4_mapped)))
    )


def _resolve_and_validate_host(hostname):
    """Resolve hostname and ensure every resolved address is public."""
    try:
        addrinfos = socket.getaddrinfo(hostname, None)
    except socket.gaierror:
        return False, None
    resolved_ips = set()
    for family, _, _, _, sockaddr in addrinfos:
        ip = sockaddr[0]
        if _is_disallowed_ip(ip):
            return False, None
        resolved_ips.add(ip)
    return True, resolved_ips


def _validate_url(url):
    """Validate a user-supplied URL to mitigate SSRF. Returns (ok, error, parsed)."""
    if not url or len(url) > 2048:
        return False, "Missing or oversized URL", None

    parsed = urlparse(url)

    if parsed.scheme.lower() not in ALLOWED_SCHEMES:
        return False, "Only http and https URLs are allowed", None

    if not parsed.hostname:
        return False, "URL must contain a valid host", None

    port = parsed.port
    if port is None:
        port = 443 if parsed.scheme.lower() == "https" else 80
    if port not in ALLOWED_PORTS:
        return False, "Port not allowed", None

    # Reject IP-literal hosts that point to internal ranges.
    try:
        literal_ip = ipaddress.ip_address(parsed.hostname)
        if _is_disallowed_ip(str(literal_ip)):
            return False, "Access to internal addresses is not allowed", None
    except ValueError:
        pass  # hostname is not an IP literal; DNS resolution check below applies.

    ok, _ = _resolve_and_validate_host(parsed.hostname)
    if not ok:
        return False, "Host resolves to a disallowed or unresolvable address", None

    return True, None, parsed


@app.route("/fetch", methods=["GET"])
def fetch_url():
    url = request.args.get("url", "").strip()

    ok, error, _ = _validate_url(url)
    if not ok:
        return jsonify({"error": error}), 400

    try:
        session = requests.Session()
        session.max_redirects = MAX_REDIRECTS

        resp = session.get(
            url,
            timeout=REQUEST_TIMEOUT,
            allow_redirects=True,
            stream=True,
        )

        # Re-validate the final URL after any redirects (redirect-based SSRF).
        final_ok, final_err, _ = _validate_url(resp.url)
        if not final_ok:
            resp.close()
            return jsonify({"error": f"Redirect target rejected: {final_err}"}), 400

        # Enforce a response size limit while reading.
        content = b""
        for chunk in resp.iter_content(chunk_size=8192):
            content += chunk
            if len(content) > MAX_CONTENT_LENGTH:
                resp.close()
                return jsonify({"error": "Response exceeds maximum allowed size"}), 413
        resp.close()

        content_type = resp.headers.get("Content-Type", "application/octet-stream")
        return Response(content, status=resp.status_code, content_type=content_type)

    except requests.exceptions.TooManyRedirects:
        return jsonify({"error": "Too many redirects"}), 400
    except requests.exceptions.Timeout:
        return jsonify({"error": "Upstream request timed out"}), 504
    except requests.exceptions.RequestException:
        return jsonify({"error": "Failed to fetch the requested URL"}), 502


if __name__ == "__main__":
    app.run(host="127.0.0.1", port=5000)